### ISSUE TITLE
GSoC 2015 mentors and project ideas

### ISSUE THREAD

**Issue by ysitu** [networkx/networkx#1341]
@hagberg @chebee7i @dschult @bjedwards and anyone who is interested

To participate in GSoC under the PSF's banner, it is necessary to have three mentors signed up for the whole summer (May 25 through Aug 24). The GSoC FAQ [suggests](https://www.google-melange.com/gsoc/document/show/gsoc_program/google/gsoc2015/help_page#howmuch) a five-hour per student per week estimate for the necessary time investment. So if we get three or four mentors on board, we can then have one or two projects. Please let me know if you can serve as mentors. Beyond project collaborators, I also hope that other contributors can lend a hand.

For project ideas, I find it mandatory to include the 2.0 API (#1246, #1336). It is arguably one of them blocker issues for 2.0. One possible project plan is to

1. fully define the API;
2. design a feasible implementation model (code complexity & performance expectation);
3. implement it (code + functional & performance tests);
4. compare with the old API (project write-up).

If this goes well, we can take the opportunity to eliminate the `_iter` methods and possibly the separate `edges`, `adjacency`, etc. functions altogether. Given its complexity, this sounds quite intense for a 12-week 40-hour-per-week commitment. If there is concern that this may not fit within 12 weeks, we can limit the scope so that it fits.

A second project idea that I can think of is an add-on/optional component system, which was discussed in #1167 and saw renewed interest in #1325. The possible project plan is similar to the one above (define, design and implement). One or two working add-ons should be included to demonstrate the system.

I have some other minor ideas (e.g., Cythonization upon installation, additional graph algorithms). Please let me know any ideas that you may have.

**Comment by chebee7i** [comment]
These all sound great. While I will be sure to chime and help out during the GSoC, I can't commit to being a mentor right now, as I'm expecting an especially busy summer. But I will definitely be around and will lend a hand!

**Comment by chebee7i** [comment]
@jtorrents @Midnighter

**Comment by chebee7i** [comment]
One other fun project could be to push on #1076.

**Comment by Midnighter** [comment]
Just to chime in quickly: This Friday I will submit my thesis, yay! So I'm definitely planning to take up work again on #1076 or could potentially mentor someone else who wants to take a stab at it. I'm not sure I qualify as a mentor, though, I've never been part of GSoC.

Another project idea: I've recently started porting some of my own projects from Python 2 to being 2&3 compatible. Packages like `six` and `future` make this quite comfortable and imho a single code base is the best way forward. I think this would go hand-in-hand with the 2.0 API and maybe a general PEP8 clean-up of `networkx`. What do you think?

**Comment by ysitu** [comment]
I need somebody to help push this forward. I am super busy these few weeks and do not have much time to deal with this.

**Comment by Midnighter** [comment]
I have some time on my hands but have not been following the details. How exactly can I help?

**Comment by ysitu** [comment]
@Midnighter The No. 1 priority is to finish up the [wiki page](https://github.com/networkx/networkx/wiki/GSoC-2015) and get three mentors to sign up. It could be too late if we do not send the proposal to the PSF within this week as Google's organization application window opens next week.

**Comment by MridulS** [comment]
Hey, I am interested in working for networkx during GSoC. How can I help?

**Comment by ysitu** [comment]
@MridulS Thank you for helping with the wiki page. Will you be able to sign up as a mentor?
@hagberg @jtorrents @dschult I really hope that at least one of you can commit to being a mentor for GSoC as you are owners of this project.

**Comment by MridulS** [comment]
@ysitu I think I don't qualify as a mentor, I am not that familiar with the codebase. But, I would love to help around during the summer or even participate as a student myself. I am an undergrad student majoring in mathematics and electrical engineering.

**Comment by MridulS** [comment]
I updated the writing your application part. I copied it from https://github.com/sympy/sympy/wiki/gsoc-2014-application-template#writing-your-proposal

**Comment by MridulS** [comment]
@ysitu How many mentors have signed up till now? According to Python Software Foundation's GSoC 2015 page we have time till March 16 to apply; the earlier the better. We should try to apply by next week.

**Comment by ysitu** [comment]
@MridulS Thanks to your help, I think that the wiki page is ready to go. The major problem is the number of mentors. PSF requires a minimum of three signed up for the summer. So far, @Midnighter has indicated that he maybe could do it. I want commit to it, but I cannot say for sure now. Then we still need one more.

**Comment by bjedwards** [comment]
If we're desperate and @hagberg, @dschult, @jtorrents, or @chebee7i can't commit, I'd be willing to be a mentor. I'm probably not the best choice as I haven't been particularly active in the last few years.

**Comment by hagberg** [comment]
You can add me to the mentor list.

**Comment by Midnighter** [comment]
@MridulS thank you for your work on the wiki page, looks great!

Just wanted to mention that for Boost.Graph there is the excellent [graph-tool](http://graph-tool.skewed.de/). 

Although there are definitely synergies with the 2.0 API, I wouldn't put using adjacency matrices strictly in the backend corner. The main appeal to me is the potential to then use powerful linear algebra on those matrices. That means that beyond a different storage format, these matrix based graphs will re-implement a number of functions that can be replaced by matrix/vector operations.

Still, I think a GSoC project for it would be great. I'd classify it as hard. I'd definitely mentor for it and I would also mentor a single source base Python 2&3 compatible implementation of the new API.

**Comment by chebee7i** [comment]
@Midnighter isn't NetworkX already a single source base for 2/3?

**Comment by Midnighter** [comment]
> isn't NetworkX already a single source base for 2/3?

True but there have been complaints about the inefficiency of `range` in Python 2, for example. Also, `dict.items` and similar are implemented as iterators for Python 2 by `six` and `future`. Nothing big but a simple `from builtins import range` at the top of most modules might already be all that's needed.

**Comment by ysitu** [comment]
I have contacted PSF's Terri Oda and am awaiting response.

@Midnighter Could you edit wiki page to clarify the descriptions of the projects that you have in mind?

**Comment by dschult** [comment]
Sorry for the late response... I'd be happy to be a mentor this summer.
I still need to finalize summer timing, but I'm hopeful that I'll have more time for NetworkX than in the past few years.

**Comment by Midnighter** [comment]
@ysitu done

**Comment by MridulS** [comment]
Having a [gitter](https://gitter.im/) chat room or an irc channel could be nice way to interact with prospective students and new developers. IMHO it is a better option (for general discussion) than github / mailing list.

P.S. I personally would prefer gitter; better interaction with github.

**Comment by gcetusic** [comment]
Hello, it's me again. I don't know if you've looked into https://github.com/espeed/bulbs. It's a good reference for support of multiple graph databases. And https://github.com/stormpath/stormpath-sdk-python/blob/master/stormpath/cache/cache.py has a nice pattern for various data storages

**Comment by rashoodkhan** [comment]
Hey!

Would it make sense to open Github issues for other GSoC project ideas as well (Alternate backends for storing graph data)?

Two project ideas are of interest to me
* Adjacency matrix based add-on - I'm more inclined on this as of my interest in graph algorithms. 
* Alternative backends for storing graph data - A very fascinating idea to work on. This seems like implementing a mini-ORM as Django provides for storing its objects. Making the backend system to be plug-able with multiple databases by providing a settings option as Django provides would be cool.

I want to discuss the above ideas with mentors/community - Where would be the best place to start the discussion, i.e. the mailing lists or Github Issues?

+1 to @MridulS  for pointing out for IRC channel for discussing ideas with mentors.

**Comment by MridulS** [comment]
@rashoodkhan Please go through #1076 to get an idea about matrix graphs.

**Comment by rashoodkhan** [comment]
@MridulS - I have started following the thread. Thanks!

--
Rashid Khan
http://www.imrashid.com

**Comment by gcetusic** [comment]
Regarding GSOC, I've written a proposal page for the NetworkX gui, first mentioned in #1326.
I didn't get any feedback on it there so please let me know what you think.
https://github.com/networkx/networkx/wiki/NetworkX-GUI---GSOC-2015-proposal

**Comment by chebee7i** [comment]
Both @bjedwards and I have visited the networkx gitter room: https://gitter.im/networkx

That seems more than fine for an IRC equivalent. No promises on occupied that room will be in general, but live conversation is often helpful.

**Comment by RONNCC** [comment]
the proposal should be much much longer - look at previous examples

**Comment by MridulS** [comment]
@chebee7i we can't access the gitter room because it is private, only owner/ collaborators can access it. Please change the settings :)

**Comment by chebee7i** [comment]
Aha. I'm guessing @hagberg do that, or create a second room that is open.

**Comment by MridulS** [comment]
Or, I guess you can add us to that room :)

**Comment by gcetusic** [comment]
@RONNCC I didn't want to spend more time on it if the idea in general is going to be considered off-topic. I asked if the idea is feasible in #1326 with no answer (so far) so I wrote the proposal anyway and @chebee7i was the first one to give a constructive criticism but hinted that although the GUI is a nice idea as Python project, the proposal might be rejected because NetworkX was leaning towards removing any GUI parts anyway. I responded with some explanations why I still consider it a good idea. Generally, I'd really like to work on NetworkX during this summer since I'm an ICT student and it's my last year so please let me know what you think of the proposal. I'm planning to make another proposal for the alternative backend.

**Comment by MridulS** [comment]
I was thinking to start writing my GSoC proposal for 2.0 API, but I have a lot of questions regarding it (specifically design). I am planning to start working on the code (I have some free time before the summer). @Midnighter I would also like to work on #1076, seems interesting. 
If we start working on 2.0 API before the summer then I think the API project won't last full summer (Will be completed by mid June). The matrix graphs project should conform with the 2.0 API. How about combining the API project with matrix graphs project?

**Comment by gcetusic** [comment]
@MridulS I'm not really that familiar with NetworkX code but any kind of access to graph data should be (in the ideal case) in one central location in the code. If that's not already the case then designing an alternative backend is going to be a nightmare and it'll change anyway during the API refactoring so these two are also clearly related as stated on the ideas page. Any ideas how coordinating these separate proposals would work?

**Comment by MridulS** [comment]
@gcetusic Yes, the design specification of 2.0 API should be done before other proposals, but there is still a lot to discuss about.

**Comment by Midnighter** [comment]
@MridulS I agree that there are definitely parallels between the different project proposals and that designing new graph classes will depend on the new API. However, I'd be cautious about taking on too much work at the same time. The new API not only includes changes to the classes but also requires adjusting the entire code base. Not always the most challenging work but definitely a lot of it. So I would wait for what applications we get and if noone applies for the matrix classes and you happen to be done early with your project then we can definitely take it on.

**Comment by MridulS** [comment]
@Midnighter But I need to explicitly mention my week by week plan in my GSoC Application. @dschult @ysitu @hagberg @chebee7i it would be helpful if we could start a discussion about the design.

**Comment by chebee7i** [comment]
@MridulS Live chat would be nice, but there is no way that all devs would be able to particpate. Given that the discussion will take place at a slower time scale and will also require time to think/reply on some of the issues, the non-threaded format of github comments and gitter is also not ideal. So my estimate is that the best place for this discussion would be on the mailing list.

That said, if you want to just chat and bounce ideas around initially, let us know...perhaps some of us might be able to join on gitter, but no guarantees (we are all quite busy folks). I've created: https://gitter.im/networkx/GSoC_2015. This room is open to all, so chat there too.

**Comment by hagberg** [comment]
Here are some other ideas for projects that are not fully scoped out.  Comments?

1. Add algorithms for directed triangles.  Clustering #1331, Triadic census #191, ...

2. Add community finding algorithms.  A previous GSoC project made some progress on this #764.  There have been other contributions since #951, #1092 (Louvain) and #1265 (Girvan-Newman), #617 (label propagation)

3.  Rework the matplotlib drawing package as a separate package from networkx #1325, with a more sensible interface, proper arrows #1248, #1357, #418.  

4.  Implement multilevel graph layout algorithms (force-directed, eigenvector)

5. Create IPython notebook based tutorials and examples 

6. Incorporate or interface to http://www.graphclasses.org/smallgraphs.html

7. Implement more shortest path algorithms (or add features) #1120.  k-shortest paths #793, restricted paths #480, etc

8. Add network metrics for weighted networks #730 

9. Add more algorithms and support for bipartite graphs #1294, #1322 

10. Better XML format readers with SAX #448 

11.  Create new web site with design and logo.  Improve documentation web site #1221

**Comment by ysitu** [comment]
@hagberg I would add the following:

* Clean up hard-coded uses of `numpy.random` and enable dependency injection a la [`std::shuffle`](http://en.cppreference.com/w/cpp/algorithm/random_shuffle) and [`java.util.Collections.shuffle`](http://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#shuffle-java.util.List-java.util.Random-).

**Comment by hagberg** [comment]
Yes, that would be fine @ysitu.  I don't think there are many places where we are using numpy.random() though.   And we do allow seeding the python random generators in many cases (though probably not all).

**Comment by MridulS** [comment]
IMO projects like 5 and 11 are not through enough to qualify as GSoC projects, but they sound as good individual projects to take on. Project 6 could help in https://github.com/networkx/networkx/issues/1343. I would personally like to work on 5, 6, 7, 8 and 9. :)

**Comment by ysitu** [comment]
@hagberg Merely allowing seeding is not enough. The underlying PRNG is shared. Seeding one function affects other functions relying on random numbers. Changing to dependency injection help decouple them and also the tests.

@MridulS 5 and 11 are about project evangelism and make a lot sense for NetworkX. I think that they are actually good projects for GSoC.

**Comment by MridulS** [comment]
Yes I agree that they are important. But I think that they won't be able to
span over a time frame of 12 weeks.

**Comment by Midnighter** [comment]
>    Add algorithms for directed triangles. Clustering #1331, Triadic census #191, ...

>    Add community finding algorithms. A previous GSoC project made some progress on this #764. There have been other contributions since #951, #1092 (Louvain) and #1265 (Girvan-Newman), #617 (label propagation)

I remember starting work on the integration but then waiting for the spectral code to be integrated. I can take a look again, modularity is certainly an important topic somewhat missing from networkx

>    Rework the matplotlib drawing package as a separate package from networkx #1325, with a more sensible interface, proper arrows #1248, #1357, #418.

>    Implement multilevel graph layout algorithms (force-directed, eigenvector)

>    Create IPython notebook based tutorials and examples

I'd love to have/create more of those. I haven't checked recent progress on https://github.com/jdfreder/ipython-d3networkx maybe @jdfreder can comment. It'd be amazing to have those visualizations available.

>    Incorporate or interface to http://www.graphclasses.org/smallgraphs.html

>    Implement more shortest path algorithms (or add features) #1120. k-shortest paths #793, restricted paths #480, etc

>    Add network metrics for weighted networks #730

>    Add more algorithms and support for bipartite graphs #1294, #1322

>    Better XML format readers with SAX #448

Although I love lxml, it's a nasty dependency to add. So I agree that SAX is the way forward.

>    Create new web site with design and logo. Improve documentation web site #1221

Not my forté but important nonetheless.

**Comment by jdfreder** [comment]
> I'd love to have/create more of those. I haven't checked recent progress on https://github.com/jdfreder/ipython-d3networkx maybe @jdfreder can comment. It'd be amazing to have those visualizations available.

It's more or less an MVP example of how one can make networkx graphs interactive in the IPython notebook.  I keep it up to date with IPython, but there's still a lot that could be done to improve what's already there.

**Comment by MridulS** [comment]
To all the mentors, mentor registrations have started http://google-melange.com/gsoc/org/list/public/google/gsoc2015

**Comment by gcetusic** [comment]
Why not just take whatever the svg on the networkx website spits out first and take it as the logo? :D

**Comment by jtorrents** [comment]
Very nice to see so much interest and cool ideas for projects! I'm very busy for the next months and will not be able to mentor, but I'll try to help as much as possibe.

**Comment by chebee7i** [comment]
@MridulS is that list final? I don't see NetworkX on it. So maybe we didn't get accepted?

**Comment by MridulS** [comment]
@chebee7i We are under Python Software Foundation, and they got in :)

**Comment by hagberg** [comment]
We are part of PSF.
https://wiki.python.org/moin/SummerOfCode/2015

**Comment by ysitu** [comment]
We are the only guy under PSF without a logo (as of now).

**Comment by hagberg** [comment]
I'd rather have proper Matplotlib arrow drawing than a logo.
But still see #11 above in my list.

**Comment by ysitu** [comment]
Time for mentor signup with Google and PSF. Announcement copy-pasted from [Soc2015-mentors](https://mail.python.org/mailman/listinfo/soc2015-mentors) mailing list:

```
[Soc2015-mentors] And we're in!

Congratulations everyone: Python is an official GSoC mentoring org for 2015!

For everyone who's going to mentor with us, it's now paperwork time:

1. Sign up on melange and request a connection with Python
2. Fill in this form so I have extra contact info for you:
http://goo.gl/forms/PMXVM1CUAS

(Melange doesn't even give me your email addresses, so that form 
generates the spreadsheet I use to keep track of you all.)

Sub org admins, please forward these two steps to anyone who will be 
mentoring with us, and you are welcome to tell folk to sign up to the 
mailing lists directly as well.  Here's the links:

https://mail.python.org/mailman/listinfo/soc2015-mentors
https://mail.python.org/mailman/listinfo/soc2015-general


  Terri
```

**Comment by prasunanand** [comment]
Hi, I just saw NetworkX is selected for GSOC 2015. Are you guys still takiang new ideas in consideration?

**Comment by hagberg** [comment]
Yes, you can discuss ideas here.

**Comment by MridulS** [comment]
@prasunanand you can also take a glance at https://github.com/networkx/networkx/wiki/GSoC-2015-project-ideas first :)

**Comment by prasunanand** [comment]
@hagberg I am currently working on integrating networkx with django and matplotlib . I basically started  it to provide tools for quick analysis of biological networks.

**Comment by prasunanand** [comment]
@MridulS  , It seems nice to see someone contributing to Networkx from BITS Goa :) :+1:

**Comment by prasunanand** [comment]
I am also going to provide graphical data analysis using javascript.

**Comment by pabloem** [comment]
Hello everyone,
I'm quite interested in working with NetworkX for this SoC. I am very interested in graph theory, and I'd like to work on some Graph algorithms. I see that #1076 is already being flirted with by other people, so I'm reading around to see if there's issues where I can help. I'll try to write a draft of a proposal soon. : )

**Comment by pabloem** [comment]
I took a look. Issue #730 seems like a project that can be done over 12 weeks - indeed, it might have a flexible size, as it just consists on expanding existing code, and possibly adding more - in the form of new measures for the network. I everybody is okay with it, I'll start preparing a proposal for this project.

@prasunanand your web application sounds like an interesting project, do you have anything up in github yet? : )

**Comment by prasunanand** [comment]
@hagberg  I didn't get any feedback. Is my idea good?

**Comment by ysitu** [comment]
@prasunanand My initial impression of your project is that it is about how one can use NetworkX for graph visualization. Do you need some development on NetworkX's side to support that? If not, that seems to be out of scope.

**Comment by unongbatai** [comment]
Hi, sorry for writing so late in the application process, can I still give my proposal now?
I'm actually interested in implementing two and eight.

> Add community finding algorithms. A previous GSoC project made some progress on this #764. There have been other contributions since #951, #1092 (Louvain) and #1265 (Girvan-Newman), #617 (label propagation)

> Add network metrics for weighted networks #730

@bjedwards can you please list out specifically which community detection algorithms you have implemented in your GSoC project? Also, what has been the cause of delay for merging them so far?

**Comment by OrkoHunter** [comment]
@unongbatai 
> Hi, sorry for writing so late in the application process, can I still give my proposal now?

The official deadline for the student proposal is 27 March 19:00 UTC. You still have time.

**Comment by ysitu** [comment]
@unongbatai If you are to write a proposal, please make sure that you provide a copy on our [wiki](https://github.com/networkx/networkx/wiki) so that everybody can take a look at it.

**Comment by MridulS** [comment]
As we are participating under the PSF umbrella, everyone should update their proposal according to the PSF template. https://wiki.python.org/moin/SummerOfCode/ApplicationTemplate2015

**Comment by OrkoHunter** [comment]
We should have one on our wiki page too. In addition to PSF's template, Networkx mentors may/should add their own expectations in it.
While I remember the earlier one (which is not now on our wiki page) was more or less a copy of SymPy's application template. We should not do that.

**Comment by unongbatai** [comment]
I won't be able to participate in GSoC this year, but I would still like to contribute.

**Comment by MridulS** [comment]
I think this can be closed now.

**Comment by hagberg** [comment]
Thanks everyone for a great GSoC.
